rust/src/x509/mod.rs

```diff
@@ -0,0 +1,104 @@
+/* Copyright (C) 2019-2020 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+// written by Pierre Chifflier  <chifflier@wzdftpd.net>
+
+use std;
+use std::ffi::CString;
+use std::os::raw::c_char;
+use x509_parser::{parse_x509_der, X509Certificate};
+
+pub struct X509(X509Certificate<'static>);
+
+#[no_mangle]
+pub extern "C" fn rs_x509_decode(input: *const u8, input_len: u32) -> *mut X509 {
+    let slice = build_slice!(input, input_len as usize);
+    let res = parse_x509_der(slice);
+    match res {
+        Ok((_rem, cert)) => {
+            Box::into_raw(Box::new(X509(cert)))
+        }
+        Err(_) => std::ptr::null_mut(),
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn rs_x509_get_subject(ptr: *const X509) -> *mut c_char {
+    if ptr.is_null() {
+        return std::ptr::null_mut();
+    }
+    let x509 = cast_pointer! {ptr, X509};
+    let subject = format!("{}", x509.0.tbs_certificate.subject);
+    let c_str = CString::new(subject).unwrap();
```

Comment:
Any chance of failure here?

Reply:
I think not (the output of `format!` is guaranteed to be valid utf-8), but I also do not like `unwrap`.
I'll rewrite that to use `to_string()` and `match` instead of `unwrap`.

Reply:
Note: This check is in fact very important: `CString::new` can fail if input data contains a null byte, which can happen in a forged certificate.

Reply:
What should happen when we have a certificate subject such as `oisf.net\x00.evil.com` ? Match only on oisf.net or on the whole buffer ?

Reply:
@catenacyber Using #4661 the conversion will fail and a null pointer is returned, then an "invalid certificate" event is raised
The existing C code silently stops at the first \0. I want to fix that, but I propose fixing that in another PR to avoid mixing changes and bugfixes